Mathlib/Data/Nat/Factorization/Multiplicity.lean

```diff
@@ -0,0 +1,302 @@
+  exact lt_of_pow_dvd_right hn p_gt h1
+
+/-- The factorization of `m` in `n` is the number of positive natural numbers `i` such that `m ^ i`
+divides `n`. Note `m` is prime. This set is expressed by filtering `Ico 1 b` where `b` is any bound
+greater than `log m n`. -/
+theorem factorization_eq_card_pow_dvd_of_lt {m n b : ℕ}
+    (hm2: m.Prime) (hn : 0 < n) (hb : log m n < b) :
+    n.factorization m = #{i ∈ Ico 1 b | m ^ i ∣ n} :=  calc
+    n.factorization m = #{i ∈ Ico 1 n | m ^ i ∣ n} := by
+      exact factorization_eq_card_pow_dvd n hm2
+    _ = #{i ∈ Ico 1 b | m ^ i ∣ n} := by
+      rw [Ico_pow_dvd_eq_Ico_of_lt hm2 (not_eq_zero_of_lt hn)]
+      exact lt_pow_of_log_lt (Prime.one_lt hm2) hb
+
+theorem factorization_pow_self {p n : ℕ} (hp : p.Prime) : (p ^ n).factorization p = n := by
+  simp [factorization_pow, Prime.factorization_self hp]
+
+/-- **Legendre's Theorem**
+
+The multiplicity of a prime in `n!` is the sum of the quotients `n / p ^ i`. This sum is expressed
+over the finset `Ico 1 b` where `b` is any bound greater than `log p n`. -/
+theorem factorization_factorial {p : ℕ} (hp : p.Prime) :
+    ∀ {n b : ℕ}, log p n < b → (n !).factorization p = (∑ i ∈ Ico 1 b, n / p ^ i : ℕ)
+  | 0, b, _ => by simp only [factorial_zero, factorization_one, coe_zero, Pi.zero_apply,
+    Nat.zero_div, sum_const_zero]
+  | n + 1, b, hb =>
+    calc
+      ((n + 1)!).factorization p = ( (n + 1) * (n !)).factorization p := by
+        rw [← factorial_succ]
+      _ = (n + 1).factorization p + (n !).factorization p := by
+        rw [factorization_mul (zero_ne_add_one n).symm (factorial_ne_zero n), coe_add, Pi.add_apply]
+      _ = #{i ∈ Ico 1 b | p ^ i ∣ n + 1} + (∑ i ∈ Ico 1 b, n / p ^ i : ℕ) := by
+        rw [factorization_factorial hp ((log_mono_right <| le_succ _).trans_lt hb)]
+        simp only [add_left_inj]
+        apply factorization_eq_card_pow_dvd_of_lt hp (zero_lt_succ n)
+        exact hb
+      _ = (∑ i ∈ Ico 1 b, n / p ^ i : ℕ) + #{i ∈ Ico 1 b | p ^ i ∣ n + 1} := by
+        exact
+          Nat.add_comm (#(Finset.filter (fun i ↦ p ^ i ∣ n + 1) (Ico 1 b)))
+            (∑ i ∈ Ico 1 b, n / p ^ i)
+      _ = (∑ i ∈ Ico 1 b, (n / p ^ i + if p ^ i ∣ n + 1 then 1 else 0) : ℕ) := by
+        rw [sum_add_distrib, sum_boole]
+        simp
+      _ = (∑ i ∈ Ico 1 b, (n + 1) / p ^ i : ℕ) :=
+        Finset.sum_congr rfl fun _ _ => (succ_div _ _).symm
+
+/-- For a prime number `p`, taking `(p - 1)` times the factorization of `p` in `n!` equals `n` minus
+the sum of base `p` digits of `n`. -/
+theorem sub_one_mul_factorization_factorial {n p : ℕ} (hp : p.Prime) :
+    (p - 1) * (n !).factorization p = n - (p.digits n).sum := by
+  simp only [factorization_factorial hp <| lt_succ_of_lt <| lt.base (log p n),
+    ← Finset.sum_Ico_add' _ 0 _ 1, Ico_zero_eq_range, ←
+    sub_one_mul_sum_log_div_pow_eq_sub_sum_digits]
+
+/-- Modified version of `factorization_prod` that accounts for inputs. -/
+theorem factorization_prod₀ {α : Type*} {p : ℕ}
```

Comment:
I'd suggest `factorization_prod_apply` as a name for this lemma. This and `factorization_pow_self` would be more at home in `Mathlib.Data.Nat.Factorization.Basic`.